File: kafka/ingestion-service/src/main/java/com/iots/kafka/ingestion/EventGenerator.java
package com.iots.kafka.ingestion;

import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;
import java.util.UUID;

/**
 * Generates sensor reading events from normalized dataset records.
 *
 * <p>Port of the shared Node.js EventGenerator: optionally restricts the dataset
 * to the first {@code deviceCount} unique devices, then emits
 * {@code totalMessages} readings by cycling through the records, attaching a
 * fresh messageId to each.</p>
 */
public class EventGenerator {

    private final List<DatasetRecord> records;

    public EventGenerator(List<DatasetRecord> records) {
        if (records == null || records.isEmpty()) {
            throw new IllegalArgumentException("EventGenerator requires at least one dataset record.");
        }
        this.records = records;
    }

    public List<SensorReading> generateEvents(int totalMessages, int deviceCount) {
        if (totalMessages < 0) {
            throw new IllegalArgumentException("totalMessages must be a non-negative integer.");
        }

        List<DatasetRecord> pool = filterByDeviceCount(deviceCount);
        if (pool.isEmpty()) {
            throw new IllegalStateException("No dataset records available for the requested controls.");
        }

        List<SensorReading> events = new ArrayList<>(totalMessages);
        for (int i = 0; i < totalMessages; i++) {
            DatasetRecord r = pool.get(i % pool.size());
            events.add(new SensorReading(
                    UUID.randomUUID().toString(),
                    r.deviceId(),
                    r.temperature(),
                    r.humidity(),
                    r.createdAt()
            ));
        }
        return events;
    }

    private List<DatasetRecord> filterByDeviceCount(int deviceCount) {
        if (deviceCount <= 0) {
            return records;
        }

        Set<String> selected = new LinkedHashSet<>();
        for (DatasetRecord r : records) {
            if (selected.size() >= deviceCount) {
                break;
            }
            selected.add(r.deviceId());
        }

        List<DatasetRecord> filtered = new ArrayList<>();
        for (DatasetRecord r : records) {
            if (selected.contains(r.deviceId())) {
                filtered.add(r);
            }
        }
        return filtered;
    }
}

File: shared/event-generator/src/eventGenerator.js
const { createSensorReading } = require("./sensorReading");

function delay(milliseconds) {
  return new Promise((resolve) => {
    setTimeout(resolve, milliseconds);
  });
}

function uniqueDeviceIds(records) {
  return Array.from(new Set(records.map((record) => record.deviceId)));
}

function filterByDeviceCount(records, deviceCount) {
  if (!deviceCount) {
    return records;
  }

  if (!Number.isInteger(deviceCount) || deviceCount < 1) {
    throw new Error("deviceCount must be a positive integer when provided.");
  }

  const selectedDevices = new Set(uniqueDeviceIds(records).slice(0, deviceCount));
  return records.filter((record) => selectedDevices.has(record.deviceId));
}

class EventGenerator {
  constructor(records) {
    if (!Array.isArray(records) || records.length === 0) {
      throw new Error("EventGenerator requires at least one normalized dataset record.");
    }

    this.records = records;
  }

  getRecordsForSimulation(options = {}) {
    const records = filterByDeviceCount(this.records, options.deviceCount);

    if (records.length === 0) {
      throw new Error("No dataset records are available for the requested simulation controls.");
    }

    return records;
  }

  generateEvents(options = {}) {
    const totalMessages = options.totalMessages;

    if (!Number.isInteger(totalMessages) || totalMessages < 0) {
      throw new Error("totalMessages must be a non-negative integer.");
    }

    const records = this.getRecordsForSimulation(options);
    const events = [];

    for (let index = 0; index < totalMessages; index += 1) {
      events.push(createSensorReading(records[index % records.length]));
    }

    return events;
  }

  async *streamEvents(options = {}) {
    const messagesPerSecond = options.messagesPerSecond;

    if (!Number.isFinite(messagesPerSecond) || messagesPerSecond <= 0) {
      throw new Error("messagesPerSecond must be a positive number.");
    }

    const records = this.getRecordsForSimulation(options);
    const intervalMs = 1000 / messagesPerSecond;
    let index = 0;

    while (true) {
      yield createSensorReading(records[index % records.length]);
      index += 1;
      await delay(intervalMs);
    }
  }
}

module.exports = {
  EventGenerator,
  filterByDeviceCount
};
